### We  change point range  0.1, 0.5, 0.8 and change rate to find the best fit for spare parts >100

### We want to get the best Prophet prediction for spare parts whose command number is >100. So we use the data from 09-17 as training data and 2018 as test data. We want to find the best  prophet paramater 'espically chang_point rate'  that can mininser the overall error

In [1]:
## import and load in
import pandas as pd
import numpy as np
from fbprophet import Prophet
import os

# select the name whose total command are >100 and save them in PARAs
df1 = pd.read_excel("encoded_pic_septembre_2018.xlsx")
df2=df1.groupby(by=['item',df1.index]).sum()
df3=df2.groupby(['item']).size()
df4 = df3.sort_values()
PARAs = df4[1521:-1].index
# >100: >1450

In [9]:
f=pd.read_csv('LRU_order_among_the_years.csv', header=1)


# delete the first line 
f=f.drop(f.index[0])
f.reset_index(drop=True, inplace=True)

# List to save every prediction errors and we can calculate their mean as overall error afterwards
error_list = []

list1=[]
list2=[]
list3=[]
worst=[]

#choose the spare part we want, delete other colomns  and change the formula to do ''Prophet'' training
for PARA in PARAs:
    df=f[['LRU',PARA]]
    df['ds'] = df['LRU']
    df['y'] = df[PARA]
    df = df.drop(['LRU',PARA],axis=1)
    ## do the sum of commandes and get the new list  
    Sum_Aberaeron = df
    Sum_Aberaeron.reset_index(drop=True, inplace=True)
    Sum_Aberaeron['ds'] = pd.to_datetime(Sum_Aberaeron['ds'])
    Sum_Aberaeron.sort_values('ds',inplace=True)
    Sum_Aberaeron['cumsum']=Sum_Aberaeron['y'].cumsum()
    Sum_Aberaeron['y']=Sum_Aberaeron['cumsum']
    Sum_Aberaeron  =Sum_Aberaeron.drop(['cumsum'],axis=1)
    
    # split the training set and test set: training :'from 2009 to '20171231' , testing :'from 20180101 to '20180926' 
    train=Sum_Aberaeron[Sum_Aberaeron['ds']<'20180101']
    test=Sum_Aberaeron[Sum_Aberaeron['ds']>'20171231']
    
    #prophet traing
    m = Prophet(growth='linear', changepoints=None, n_changepoints=25, changepoint_range=0.5, yearly_seasonality='auto', weekly_seasonality='auto', daily_seasonality='auto', holidays=None, seasonality_mode='additive', seasonality_prior_scale=10.0, holidays_prior_scale=10.0, changepoint_prior_scale=0.05, mcmc_samples=0, interval_width=0.8, uncertainty_samples=1000)
    m.fit(train)
    forecast = m.predict(test)
    
 # calculate the grow rate in year 2018  for both the true data and prediction
    firstday = forecast[forecast['ds']=='20180101']
    lastday = forecast[forecast['ds']=='20180926']
    agg_2018_pre= (lastday['yhat'].values[0] - firstday['yhat'].values[0])
    #print(PARA+' predicted command number in 2018 : ' +str(agg_2018_pre))

    # print("Generalization error: %g"%(1.-rf.score(Xtest,ytest)))

    firstday_test = test[test['ds']=='20180101']
    lastday_test = test[test['ds']=='20180926']
    agg_2018 = (lastday_test['y'].values[0] - firstday_test['y'].values[0])
    #print(PARA+' true command number in 2018: ' + str(agg_2018))

    #calculate the prediction error
    error=np.abs(agg_2018-agg_2018_pre)/agg_2018 
    if error<0.1:
        list1.append(PARA)
    else:
        if error<0.4:
            list2.append(PARA)
        else:
            if error<0.9:
                list3.append(PARA)
            else:
                worst.append(PARA)
    print(PARA+' PREDICT_error: ' + str(error ))
    error_list.append(error)

/Users/qiuqiuli/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
/Users/qiuqiuli/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.
INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Chester PREDICT_error: 0.25443951874661414


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Buellton PREDICT_error: 0.6408018892035009


/Users/qiuqiuli/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:54: RuntimeWarning: divide by zero encountered in double_scalars
INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cloquet PREDICT_error: inf


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Florham Park PREDICT_error: 0.958676511709333


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


College Park PREDICT_error: 29.471057271261273


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Darien PREDICT_error: 0.6218377570633579


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Gastonia PREDICT_error: 0.09746337911566785


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Grayson PREDICT_error: 0.49980392726424977


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Glendora PREDICT_error: 0.010537870375066214


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Galax PREDICT_error: 0.41706257981292366


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Farmington PREDICT_error: 0.4418920663226927


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Belvidere PREDICT_error: 0.09420428913519703


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Fairfield PREDICT_error: 1.0002366892473902


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Dardenne Prairie PREDICT_error: 0.6688370767881829


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


El Paso PREDICT_error: 0.23626607698255403


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Fort Bragg PREDICT_error: 0.18489173682973137


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


El Mirage PREDICT_error: 0.105506466400625


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Glenolden PREDICT_error: 1.814193332550983


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Glens Falls PREDICT_error: 0.31413114985707374


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Grosse Pointe Park PREDICT_error: 0.6923208552064921


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cahokia PREDICT_error: 0.5063087545378534


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cameron PREDICT_error: 0.2074094189641086


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Canfield PREDICT_error: 0.053760369640860305


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Benton PREDICT_error: inf


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Camden Town PREDICT_error: 0.3853617308308474


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Calera PREDICT_error: 0.3268262280021697


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cedarburg PREDICT_error: 0.4215631566005854


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Canal Fulton PREDICT_error: 0.8058655720955872


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Bay Harbor Islands PREDICT_error: 0.06513675084875226


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Elkhart PREDICT_error: 0.22011840641202954


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cheney PREDICT_error: 0.5272039741937116


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


El Dorado PREDICT_error: 0.025801684623792


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Catasauqua PREDICT_error: 0.6895581654773387


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


El Cerrito PREDICT_error: 0.028788917421913744


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Elberton PREDICT_error: 1.0625190035613308


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cedar Hill PREDICT_error: 0.048401528824346556


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Farmersville PREDICT_error: 0.06002581364481521


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Farmville PREDICT_error: 0.09275186031203238


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Collegeville PREDICT_error: 0.7037648122076031


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Farmington Hills PREDICT_error: 0.13824475502125663


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Elburn PREDICT_error: 0.5139219722203928


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Collingdale PREDICT_error: 0.5019026805744427


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cedar Falls PREDICT_error: 1.18231792727266


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Catford PREDICT_error: 6.399719302732649


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Bennettsville PREDICT_error: inf


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cheraw PREDICT_error: 0.40421128234613635


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cave Creek PREDICT_error: 1.6516937088720927


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Crockett PREDICT_error: 0.6055296262683302
Eldridge PREDICT_error: 0.14823789688282007


In [6]:
error_list.sort()
error_list.pop(-1)
#error_list.pop(-1)
#error_list.pop(-1)
error_list

[0.010537870375066214,
 0.025801684623792,
 0.028788917421913744,
 0.048401528824346556,
 0.053760369640860305,
 0.06002581364481521,
 0.06513675084875226,
 0.09275186031203238,
 0.09420428913519703,
 0.09746337911566785,
 0.105506466400625,
 0.13824475502125663,
 0.14823789688282007,
 0.18489173682973137,
 0.2074094189641086,
 0.22011840641202954,
 0.23626607698255403,
 0.25443951874661414,
 0.31413114985707374,
 0.3268262280021697,
 0.3853617308308474,
 0.40421128234613635,
 0.41706257981292366,
 0.4215631566005854,
 0.4418920663226927,
 0.49980392726424977,
 0.5019026805744427,
 0.5063087545378534,
 0.5139219722203928,
 0.5272039741937116,
 0.6055296262683302,
 0.6218377570633579,
 0.6408018892035009,
 0.6688370767881829,
 0.6895581654773387,
 0.6923208552064921,
 0.7037648122076031,
 0.8058655720955872,
 0.958676511709333,
 1.0002366892473902,
 1.0625190035613308,
 1.18231792727266,
 1.6516937088720927,
 1.814193332550983,
 6.399719302732649]

In [14]:
list1

['Gastonia',
 'Glendora',
 'Belvidere',
 'Canfield',
 'Bay Harbor Islands',
 'El Dorado',
 'El Cerrito',
 'Cedar Hill',
 'Farmersville',
 'Farmville']

In [15]:
list2

['Chester',
 'El Paso',
 'Fort Bragg',
 'El Mirage',
 'Glens Falls',
 'Cameron',
 'Camden Town',
 'Calera',
 'Elkhart',
 'Farmington Hills',
 'Eldridge']

In [16]:
list3

['Buellton',
 'Darien',
 'Grayson',
 'Galax',
 'Farmington',
 'Dardenne Prairie',
 'Grosse Pointe Park',
 'Cahokia',
 'Cedarburg',
 'Canal Fulton',
 'Cheney',
 'Catasauqua',
 'Collegeville',
 'Elburn',
 'Collingdale',
 'Cheraw',
 'Crockett']

In [39]:
f=pd.read_csv('LRU_order_among_the_years.csv', header=1)


# delete the first line 
f=f.drop(f.index[0])
f.reset_index(drop=True, inplace=True)

# List to save every prediction errors and we can calculate their mean as overall error afterwards
error_list = []

#choose the spare part we want, delete other colomns  and change the formula to do ''Prophet'' training
for PARA in list3:
    df=f[['LRU',PARA]]
    df['ds'] = df['LRU']
    df['y'] = df[PARA]
    df = df.drop(['LRU',PARA],axis=1)
    ## do the sum of commandes and get the new list  
    Sum_Aberaeron = df
    Sum_Aberaeron.reset_index(drop=True, inplace=True)
    Sum_Aberaeron['ds'] = pd.to_datetime(Sum_Aberaeron['ds'])
    Sum_Aberaeron.sort_values('ds',inplace=True)
    Sum_Aberaeron['cumsum']=Sum_Aberaeron['y'].cumsum()
    Sum_Aberaeron['y']=Sum_Aberaeron['cumsum']
    Sum_Aberaeron  =Sum_Aberaeron.drop(['cumsum'],axis=1)
    
    # split the training set and test set: training :'from 2009 to '20171231' , testing :'from 20180101 to '20180926' 
    train=Sum_Aberaeron[Sum_Aberaeron['ds']<'20180101']
    test=Sum_Aberaeron[Sum_Aberaeron['ds']>'20171231']
    
    #prophet traing
    m = Prophet(growth='linear', changepoints=None, n_changepoints=25, changepoint_range=0.8, yearly_seasonality='auto', weekly_seasonality='auto', daily_seasonality='auto', holidays=None, seasonality_mode='additive', seasonality_prior_scale=10.0, holidays_prior_scale=10.0, changepoint_prior_scale=0.05, mcmc_samples=0, interval_width=0.8, uncertainty_samples=1000)
    m.fit(train)
    forecast = m.predict(test)
    
 # calculate the grow rate in year 2018  for both the true data and prediction
    firstday = forecast[forecast['ds']=='20180101']
    lastday = forecast[forecast['ds']=='20180926']
    agg_2018_pre= (lastday['yhat'].values[0] - firstday['yhat'].values[0])
    #print(PARA+' predicted command number in 2018 : ' +str(agg_2018_pre))

    # print("Generalization error: %g"%(1.-rf.score(Xtest,ytest)))

    firstday_test = test[test['ds']=='20180101']
    lastday_test = test[test['ds']=='20180926']
    agg_2018 = (lastday_test['y'].values[0] - firstday_test['y'].values[0])
    #print(PARA+' true command number in 2018: ' + str(agg_2018))

    #calculate the prediction error
    error=np.abs(agg_2018-agg_2018_pre)/agg_2018 
    print(PARA+' PREDICT_error: ' + str(error ))
    error_list.append(error)

/Users/qiuqiuli/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Users/qiuqiuli/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.
INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Buellton PREDICT_error: 0.4966603785756574


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Darien PREDICT_error: 0.618723770349502


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Grayson PREDICT_error: 0.7463021571672758


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Galax PREDICT_error: 0.48276290032228847


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Farmington PREDICT_error: 0.6066970114195391


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Dardenne Prairie PREDICT_error: 0.6864524507171631


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Grosse Pointe Park PREDICT_error: 0.2854176342779843


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cahokia PREDICT_error: 0.4557577981671992


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cedarburg PREDICT_error: 0.12151573196907965


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Canal Fulton PREDICT_error: 0.5890545810213003


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cheney PREDICT_error: 0.07333608284820678


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Catasauqua PREDICT_error: 0.27634810562998924


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Collegeville PREDICT_error: 0.5989064957757978


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Elburn PREDICT_error: 0.5502708627235531


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Collingdale PREDICT_error: 0.16440899462056333


INFO:fbprophet.forecaster:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


Cheraw PREDICT_error: 0.13979265238615696
Crockett PREDICT_error: 0.6064349051776904


In [40]:
print(np.mean(error_list))
print(np.std(error_list))

0.4411083831264086
0.21136980889134618


### list1
0.1: mean 0.51 std 0.07

0.5: mean 0.058 std 0.029

0.8: mean 0.13 std 0.11

### list2
0.1: mean 0.54 std 0.1

0.5: mean 0.23 std 0.08

0.8: mean 0.27 std 0.15

### list3
0.1: mean 0.64 std 0.26

0.5: mean 0.57 std 0.12

0.8: mean 0.51 std 0.28

1.0: mean 